aioredlock/redis.py

```diff
@@ -400,7 +402,8 @@ async def unset_lock(self, resource, lock_identifier):
                        resource, successful_removes, len(self.instances), elapsed_time)
 
         if not unlocked:
-            raise LockError('Can not release the lock')
+            errors = [e for e in successes if isinstance(e, Exception)] or [None]
+            raise LockError('Can not release the lock') from errors[0]
```

Comment:
Unfortunately, raising from [None] is going to raise an undesired effect I think
```
>>> raise Exception('this') from [None]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: exception causes must derive from BaseException
```
Shouldn't there be an exception tin errors if it was unable to lock?  Why do we need the or statement.

Reply:
If this is not the case, can we write a test that enforces this situation, where the [None] would be used?

Reply:
Hi @gtmanfred !

> Unfortunately, raising from [None] is going to raise an undesired effect I think

The code won't raise from `[None]` but from `None` (because of `errors[0]` and `[None][0] = None`).

> Shouldn't there be an exception tin errors if it was unable to lock? Why do we need the or statement.

In the previous version, I was ignoring `asyncio.CancelledError` (the only error that don't inherit from `Exception` but from `BaseException`) and stopping the chain (`raise ... from None`) for them.
But that was a bad idea and I've updated the code to chain `asyncio.CancelledError`s as well with:
```diff
--- a/aioredlock/redis.py
+++ b/aioredlock/redis.py
@@ -342,7 +342,7 @@ class Redis:
                        resource, successful_sets, len(self.instances), elapsed_time)
 
         if not locked:
-            errors = [e for e in successes if isinstance(e, Exception)] or [None]
+            errors = [e for e in successes if isinstance(e, BaseException)]
             raise LockError('Can not acquire the lock "%s"' % resource) from errors[0]
 
         return elapsed_time
@@ -372,7 +372,7 @@ class Redis:
                        resource, len(successful_list), len(self.instances), elapsed_time)
 
         if not success:
-            errors = [e for e in successes if isinstance(e, Exception)] or [None]
+            errors = [e for e in successes if isinstance(e, BaseException)]
             raise LockError('Could not fetch the TTL for lock "%s"' % resource) from errors[0]
 
         return successful_list[0]
```
But there's a catch un `unset_lock` because it can be called with `ensure_future` if we get a `CancelledError` in `set_lock`:
https://github.com/joanvila/aioredlock/blob/49dc1aff64c50bc59c17735d09a2922e03171f1c/aioredlock/algorithm.py#L96-L101

In this case, `if not unlocked:` is `True` (because `successes` list is empty) and if don't have a fallback value (`or [None]`) for exception chaining then we get a list index error.
What do you think?
(PS: I'm not an expert and I may have missed something so don't hesitate to tell!)

Reply:
Is it possible to write a test for this edge case?

Reply:
Is it possible to write a test for this edge case?

Reply:
I would have love to but when coming back to this issue I finally understood what was happening here..

As I showed in the updated examples, I am using this code structure with aioredlock:
```py
try:
  lock = await lock_manager.lock()
except LockError as e:
  ... log error or raise
else:
  ... if no exception, run the code protected by the lock
  await lock_manager.unlock(lock)
finally:
  await lock_manager.destroy()
```
In this case, the `finally` block will be executed whether or not there was an error when acquiring the lock.
But if we get an exception in `set_lock` then the code in `algorithm.py` will catch it and schedule a task to cleanup later:
https://github.com/joanvila/aioredlock/blob/49dc1aff64c50bc59c17735d09a2922e03171f1c/aioredlock/algorithm.py#L96-L101
It conflicts with the `destroy` we do ourselves which itself calls `unset_lock` and destroys our `Instance` objects.
When the scheduled task is executed, there is no instances to clean anymore as they were destroyed and with the conditions in `unset_lock` we end up here:
https://github.com/joanvila/aioredlock/blob/49dc1aff64c50bc59c17735d09a2922e03171f1c/aioredlock/redis.py#L403

That's why I had to protect the `raise LockError(...) from [None][0]` in my previous comment.
When not using `finally` in my code then no problem anymore and I can remove the default value to raise from.

Unfortunately this reveals an issue with the current Aioredlock code that need to be fixed in another PR: if there is an exception on one of the instance when setting a lock then resources acquired on other instances are never cleaned up. Not cool.

Anyway, I've pushed a new version of the PR.

Reply:
Perfect, thanks for the detail.  This looks good now. Lets get it merged, and then we can address the other issue.